torch/fx/passes/dialect/common/cse_pass.py

```diff
@@ -0,0 +1,86 @@
+        """
+        Return a new copy of torch.fx.GraphModule with CSE applied to the input graph
+        
+        Example usage:
+
+        from functorch import make_fx
+        def f(a):
+            b = a * a
+            c = a * a
+            return b+c
+
+        p = CSEPass()
+        traced_graph = make_fx(f)(torch.tensor(1))
+        print(traced_graph)
+        result = p(traced_graph)
+        print(result.graph_module)
+        """
+        new_graph = fx.Graph()
+        env = {}  # map from node in the old graph to node in the new graph
+        hash_env = {}  # map from hash to a node in the new graph
+        token_map = {}  # map from hash to token
+        for n in graph_module.graph.nodes:
+            # The placeholder, output, and get_attr nodes are copied to the new grpah without change
+            # do not CSE away random operations
+            if n.op == 'placeholder' or n.op == 'output' or n.op == 'get_attr' or get_aten_target(n) in rand_ops:
+                new_node = new_graph.node_copy(n, lambda x: env[x])
+                env[n] = new_node
+            else:  # n.op == 'call_function', should never see n.op == 'call_module' or 'call_method'
+                # substitute args and kwargs memebrs to their mapping in env if exists
+                # specs can be used to reconstruct nested list/dictionaries
+                def substitute(arg_list):
+                    arg_list, spec = tree_flatten(arg_list)
+                    for i in range(len(arg_list)):
+                        v = arg_list[i]
+                        if isinstance(v, torch.fx.node.Node) and v in env:
+                            arg_list[i] = env[v]
+                    return tuple(arg_list), spec
+                args, args_spec = substitute(n.args)
+                kwargs, kwargs_spec = substitute(n.kwargs)
+
+                # each token corresponds to a unique node
+                # nodes with the same token can be substituted
+                token = {"target": n.target, "args": args, "args_spec": args_spec,
+                        "kwargs": kwargs, "kwargs_spec": kwargs_spec}
+
+                # hash substituted args to a number, do not hash specs because specs are not hashable
+                hash_arg = hash((args, kwargs))
+                hash_val = (n.target, hash_arg)
+
+                # check if a node has a substitute and can be eliminated
+                hash_val_in_hash_env = hash_val in hash_env
+                if hash_val_in_hash_env and token_map[hash_val] == token:
+                    env[n] = hash_env[hash_val]
+                    continue
+
+                new_node = new_graph.node_copy(n, lambda x: env[x])
+                env[n] = new_node
+                if not hash_val_in_hash_env:
+                    hash_env[hash_val] = new_node
+                    token_map[hash_val] = token
+
+        csed_gm =  fx.GraphModule(graph_module, new_graph)
+        return PassResult(csed_gm, False)
```

Comment:
The second argument in PassResult is a flag that is set to `True` if you have modified the graph. From a quick skim of your code I think you would just set it to `True` if it ever gets to the end of your else statement. Do you think you could incorporate this into your pass (it doesn't have to be in this PR)?

Reply:
oh I don't think I modified the graph? At the beginning there's `new_graph = fx.Graph()`, and we just copy nodes from the node graph to this new graph.

Reply:
mm by "if you have modified the graph" I mean that it represents if the graph returned is different than the graph inputted. For context, we want to keep running passes until the graphs converge -- so if the new graph created is the same as the graph inputted, that means the graph has converged so `modified = False` and we stop looping.

Reply:
oh I see, makes sense! fixed now. When a substitution happens, and less node is copied, it returns modified=True

Reply:
Thanks! lgtm 😍